File: bindings/dotnet/AIBridge/Hello.cs
namespace AIBridge;

// ============================================================================
// Hello world 示例（Program Main）
//
// 验证 .NET 绑定全链路：echo 适配器免认证可用
//   1. new Client("echo") + Start
//   2. Chat(echo-chat, [user:hello]) → 打印 choices[0].message.content（期望 "hello [echo]"）
//   3. ChatStreamAsync 同上 → await foreach 打印 chunk（期望 3 个）
//   4. Speech(echo-tts, input:hello, voice:alloy) → 打印 AudioData.Length（期望 15）
//   5. Dispose
//
// 运行方式：dotnet run（需先 cargo build -p aibridge-ffi 产 libaibridge.dylib）
// ============================================================================

public static class Hello
{
    public static async Task<int> Main(string[] args)
    {
        Console.WriteLine("=== AIBridge .NET Hello World (echo adapter) ===\n");

        try
        {
            // 1. 创建 + 启动客户端（echo 免认证）
            using var client = new Client("echo");
            client.Start();
            Console.WriteLine("[OK] Client(echo) created and started.\n");

            // 2. Chat：echo 回显 +" [echo]"
            Console.WriteLine("--- Chat ---");
            var chatReq = new ChatRequest("echo-chat", new[]
            {
                ChatMessage.User("hello"),
            });
            ChatCompletion completion = client.Chat(chatReq);
            string? content = completion.Choices.Count > 0
                ? completion.Choices[0].Message.Content
                : null;
            Console.WriteLine($"  choices[0].message.content = {content ?? "(空)"}");
            Console.WriteLine($"  期望: \"hello [echo]\"，实际: \"{content}\"");
            Console.WriteLine($"  匹配: {content == "hello [echo]"}\n");

            // 3. ChatStream：echo 返 3 个 chunk
            Console.WriteLine("--- ChatStream ---");
            var streamReq = new ChatRequest("echo-chat", new[]
            {
                ChatMessage.User("hello"),
            });
            int chunkCount = 0;
            var assembled = new System.Text.StringBuilder();
            await foreach (ChatCompletionChunk chunk in client.ChatStreamAsync(streamReq))
            {
                chunkCount++;
                string? delta = chunk.Choices.Count > 0
                    ? chunk.Choices[0].Delta.Content
                    : null;
                Console.WriteLine($"  chunk #{chunkCount}: delta.content={delta ?? "(空)"}");
                if (delta != null) assembled.Append(delta);
            }
            Console.WriteLine($"  共 {chunkCount} 个 chunk，拼接结果: \"{assembled}\"");
            Console.WriteLine($"  期望 3 个 chunk，实际: {chunkCount}\n");

            // 4. Speech：echo 返 15 字节固定音频
            Console.WriteLine("--- Speech ---");
            var speechReq = new SpeechRequest("echo-tts", "hello", "alloy");
            SpeechResult speech = client.Speech(speechReq);
            Console.WriteLine($"  AudioData.Length = {speech.AudioData.Length}");
            Console.WriteLine($"  format = {speech.Format}, model = {speech.Model}");
            Console.WriteLine($"  期望 15 字节，实际: {speech.AudioData.Length}\n");

            // 5. Dispose（using 自动调）
            Console.WriteLine("[OK] Client disposed.");

            // 汇总
            Console.WriteLine("\n=== 汇总 ===");
            Console.WriteLine($"  Chat 回显匹配: {content == "hello [echo]"}");
            Console.WriteLine($"  Stream chunk 数匹配(==3): {chunkCount == 3}");
            Console.WriteLine($"  Speech 字节数匹配(==15): {speech.AudioData.Length == 15}");
            return 0;
        }
        catch (AibridgeException ex)
        {
            Console.Error.WriteLine($"[FAIL] AibridgeException: code={ex.Code}, retryable={ex.Retryable}");
            Console.Error.WriteLine($"  message: {ex.Message}");
            return 1;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[FAIL] {ex.GetType().Name}: {ex.Message}");
            Console.Error.WriteLine(ex);
            return 2;
        }
    }
}

File: bindings/go/example/hello.go
// AIBridge Go 绑定 - Hello World 示例
//
// 使用 echo adapter（provider="echo"，免认证）验证跨语言管线：
//   - Chat：回显最后一条 user 消息 + " [echo]"
//   - ChatStream：3 个 chunk（role / 前半段 / 后半段+finish）
//   - Speech：返回 15 字节 mock 音频（b"mock-audio-data"）
//
// 运行：
//
//	cd bindings/go
//	CGO_ENABLED=1 go run ./example
//
// 需先 cargo build -p aibridge-ffi 产出 target/debug/libaibridge.dylib。
// macOS 若报 dylib 找不到，设置：
//
//	DYLD_LIBRARY_PATH=/Users/skywing/agn-sdk/target/debug
package main

import (
	"fmt"
	"os"

	aibridge "github.com/aibridge/aibridge-go"
)

func main() {
	fmt.Println("=== AIBridge Go 绑定 Hello World ===")
	fmt.Println()

	// 1. 创建并启动 echo 客户端（免认证）
	client, err := aibridge.NewClient("echo", nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "NewClient 失败: %v\n", err)
		os.Exit(1)
	}
	defer client.Close() // RAII：确保释放句柄

	if err := client.Start(); err != nil {
		fmt.Fprintf(os.Stderr, "Start 失败: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("[OK] echo 客户端已创建并启动")
	fmt.Println()

	// 2. Chat：回显 hello + " [echo]"
	fmt.Println("--- Chat ---")
	chatReq := &aibridge.ChatRequest{
		Model: "echo-chat",
		Messages: []aibridge.ChatMessage{
			aibridge.NewUserTextMessage("hello"),
		},
	}
	completion, err := client.Chat(chatReq)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Chat 失败: %v\n", err)
		os.Exit(1)
	}
	content := ""
	if len(completion.Choices) > 0 {
		content = completion.Choices[0].Message.Content
	}
	fmt.Printf("[OK] Chat 返回: id=%s, model=%s\n", completion.ID, completion.Model)
	fmt.Printf("     choices[0].message.content = %q (期望 %q)\n", content, "hello [echo]")
	if content != "hello [echo]" {
		fmt.Fprintf(os.Stderr, "     [FAIL] 内容不匹配\n")
		os.Exit(1)
	}
	fmt.Println()

	// 3. ChatStream：3 个 chunk
	fmt.Println("--- ChatStream ---")
	streamReq := &aibridge.ChatRequest{
		Model: "echo-chat",
		Messages: []aibridge.ChatMessage{
			aibridge.NewUserTextMessage("hello"),
		},
	}
	stream, err := client.ChatStream(streamReq)
	if err != nil {
		fmt.Fprintf(os.Stderr, "ChatStream 失败: %v\n", err)
		os.Exit(1)
	}

	chunkCount := 0
	var assembledContent string
	var finishReason string
	for chunk := range stream.Ch() {
		chunkCount++
		if len(chunk.Choices) > 0 {
			delta := chunk.Choices[0].Delta
			if delta.Role != "" {
				fmt.Printf("     chunk %d: role=%q\n", chunkCount, delta.Role)
			}
			if delta.Content != "" {
				fmt.Printf("     chunk %d: content=%q\n", chunkCount, delta.Content)
				assembledContent += delta.Content
			}
			if chunk.Choices[0].FinishReason != "" {
				finishReason = chunk.Choices[0].FinishReason
				fmt.Printf("     chunk %d: finish_reason=%q\n", chunkCount, finishReason)
			}
		}
	}
	if err := stream.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "     [FAIL] 流式错误: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("[OK] ChatStream 收到 %d 个 chunk (期望 3)\n", chunkCount)
	fmt.Printf("     拼接内容 = %q (期望 %q)\n", assembledContent, "hello [echo]")
	fmt.Printf("     finish_reason = %q\n", finishReason)
	if chunkCount != 3 {
		fmt.Fprintf(os.Stderr, "     [FAIL] chunk 数量不匹配\n")
		os.Exit(1)
	}
	fmt.Println()

	// 4. Speech：15 字节 mock 音频
	fmt.Println("--- Speech ---")
	speechReq := &aibridge.SpeechRequest{
		Model: "echo-tts",
		Input: "hello",
		Voice: aibridge.SingleVoice("alloy"),
	}
	speech, err := client.Speech(speechReq)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Speech 失败: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("[OK] Speech 返回: model=%s, format=%s, content_type=%s\n",
		speech.Model, speech.Format, speech.ContentType)
	fmt.Printf("     len(AudioData) = %d (期望 15)\n", len(speech.AudioData))
	fmt.Printf("     AudioData = %q\n", string(speech.AudioData))
	if len(speech.AudioData) != 15 {
		fmt.Fprintf(os.Stderr, "     [FAIL] 音频字节数不匹配\n")
		os.Exit(1)
	}
	fmt.Println()

	fmt.Println("=== 全部通过 ===")
}
